services/server/server.go

```diff
@@ -110,89 +110,44 @@ func (s *Server) UpdateSettingsFromEnv(env []string) error {
 	s.envRW.Lock()
 	defer s.envRW.Unlock()
 
-	return s.db.InTransaction(func(tx *reform.TX) error {
-		settings, err := models.GetSettings(tx.Querier)
-		if err != nil {
-			return err
-		}
-
-		for _, e := range env {
-			p := strings.SplitN(e, "=", 2)
-			if len(p) != 2 {
-				s.l.Warnf("Failed to parse environment variable %q.", e)
-				continue
-			}
+	envSettings, errs, warns := envvars.ParseEnvVars(env)
+	for _, w := range warns {
+		s.l.Warnln(w)
+	}
 
-			k, v := strings.ToUpper(p[0]), strings.ToLower(p[1])
-			var err error
-			var value interface{}
-			switch k {
-			case "DISABLE_UPDATES":
-				var b bool
-				b, err = strconv.ParseBool(v)
-				if err == nil {
-					value = b
-					s.envDisableUpdates = b
-				}
-
-			case "DISABLE_TELEMETRY":
-				var b bool
-				b, err = strconv.ParseBool(v)
-				if err == nil {
-					value = b
-					s.envDisableTelemetry = b
-					settings.Telemetry.Disabled = b
-				}
-
-			case "METRICS_RESOLUTION", "METRICS_RESOLUTION_HR":
-				var d time.Duration
-				d, err = time.ParseDuration(v)
-				if err == nil {
-					value = d
-					s.envMetricsResolutionHR = d
-					settings.MetricsResolutions.HR = d
-				}
-
-			case "METRICS_RESOLUTION_MR":
-				var d time.Duration
-				d, err = time.ParseDuration(v)
-				if err == nil {
-					value = d
-					s.envMetricsResolutionMR = d
-					settings.MetricsResolutions.MR = d
-				}
-
-			case "METRICS_RESOLUTION_LR":
-				var d time.Duration
-				d, err = time.ParseDuration(v)
-				if err == nil {
-					value = d
-					s.envMetricsResolutionLR = d
-					settings.MetricsResolutions.LR = d
-				}
-
-			case "DATA_RETENTION":
-				var d time.Duration
-				d, err = time.ParseDuration(v)
-				if err == nil {
-					value = d
-					s.envDataRetention = d
-					settings.DataRetention = d
-				}
+	// This should be impossible in the normal workflow.
+	// An invalid environment variable must be caught with pmm-managed-init
+	// and the docker run must be terminated.
+	if len(errs) > 0 {
+		for _, e := range errs {
+			s.l.Errorln(e)
+		}
+		return errors.New("validation error of environment variables")
+	}
 
-			default:
-				s.l.Tracef("Skipping %q.", e)
-				continue
-			}
+	s.envDisableUpdates = envSettings.DisableUpdates
+	s.envDisableTelemetry = envSettings.DisableTelemetry
+	s.envMetricsResolutionHR = envSettings.MetricsResolutions.HR
+	s.envMetricsResolutionMR = envSettings.MetricsResolutions.MR
+	s.envMetricsResolutionLR = envSettings.MetricsResolutions.LR
+	s.envDataRetention = envSettings.DataRetention
```

Comment:
If ParseEnvVars returns `*ChangeSettingsParams`, then we can just save them as a single field.